File: leetcode6305.js
/*
* auther yeling
* 
* 
*/
var isPossibleToCutPath = function (grid) {
    let row = grid.length;
    let col = grid[0].length;
    let n = row * col;
    let res = [];
    let dfn = new Array(n).fill(-1);
    let low = new Array(n).fill(n);
    let num = 0;

    connections = []
    for(let i = 0; i < row; i++) {
        for(let j = 0; j < col; j++) {
            if(grid[i][j] == 1) {
                if(i + 1 < row && grid[i + 1][j] == 1) {
                    connections.push([i*row + j, (i+1)*row + j])
                }
                if(j + 1 < col && grid[i][j + 1] == 1) {
                    connections.push([i*row + j, i*row + j + 1])
                }
            }
        }
    }


    let dfs = (allPos, index, father) => {
        // console.log(`dfs ${index}`);
        if (dfn[index] == -1) {
            dfn[index] = num++;
            low[index] = dfn[index];
        }
        let childs = allPos[index];
        for (let i = 0; i < childs.length; i++) {
            if (low[childs[i]] == n) {
                dfs(allPos, childs[i], index);
                low[index] = Math.min(low[index], low[childs[i]]);
            } else if (childs[i] != father) {
                low[index] = Math.min(low[index], low[childs[i]]);
            }
        }
        // console.log(dfn);
        // console.log(low);
        return low[index]

    }
    let allPos = new Array(n).fill(0).map(() => []);
    for (let i = 0; i < connections.length; i++) {
        allPos[connections[i][0]].push(connections[i][1]);
        allPos[connections[i][1]].push(connections[i][0]);
    }
    // console.log(allPos);
    dfs(allPos, 0, -1);
    // console.log(dfn);
    // console.log(low);

    for (let i = 0; i < connections.length; i++) {
        let curr = connections[i];
        if (dfn[curr[0]] < low[curr[1]] || dfn[curr[1]] < low[curr[0]]) {
            res.push(curr);
        }
    }
    
    console.log(res);

};

// grid = [[1, 1, 1], [1, 0, 1], [1, 1, 1]]
grid = [[1,1,1],[1,0,0],[1,1,1]]
console.log(isPossibleToCutPath(grid))

File: leetcode6305.py

# auther yeling
from typing import List
from bisect import *
from collections import *
from functools import *
from itertools import *
from math import *
from queue import PriorityQueue

INF = 2 ** 64 - 1
MOD = 10 ** 9 + 7

class Solution:
    def isPossibleToCutPath(self, grid: List[List[int]]) -> bool:
        #查找father
        def find(father, u):
            if father[u] != u:
                father[u] = find(father,father[u])
            return father[u]
        #合并
        def join(father, u, v):
            fu = find(father,u)
            fv = find(father,v)
            if fu != fv:
                father[fu] = fv
        m = len(grid)
        n = len(grid[0])
        father = list(range(m*n))
        for i in range(m):
             for j in range(n):
                if i == 0 and j == 0:
                    continue
                if grid[i][j] == 1 and i + 1 < m and grid[i + 1][j] == 1:
                    join(father, i*m + j, (i + 1) * j)
                if grid[i][j] == 1 and j + 1 < n and grid[i][j + 1] == 1:
                    join(father, i*m + j, i * (j + 1))
        
        if find(father,1) == find(father, m*n - 1) and find(father,n) == find(find,m*n - 1):
            return False
        else:
            return True
        return
    
grid = [[1,1,1],[1,0,0],[1,1,1]]
grid = [[1,1,1],[1,0,1],[1,1,1]]
sol = Solution()
print(sol.isPossibleToCutPath(grid))
